lib/export.js

```diff
@@ -2400,6 +2411,66 @@ class FHIRExporter {
     }
   }
 
+  applyOwnSubsetConstraintsOnExtension(sourceValue, profile, snapshotEl, differentialEl) {
+    const subsetConstraints = sourceValue.constraintsFilter.own.subset.constraints;
+
+    if (subsetConstraints.length > 0) {
+      [snapshotEl, differentialEl] = this.findConstrainableElement(sourceValue, profile, snapshotEl, differentialEl);
+
+      const sourceIdentifiers = common.subsetEffectiveIdentifiers(sourceValue);
+      const targetTypes = snapshotEl.type;
+
+      // Note that this code is very similar to code in applyOwnTypeConstraintsOnExtension, but this code
+      // loops over multiple possible sourceIdentifiers in a loop instead of acting on only one
+      for (const sourceIdentifier of sourceIdentifiers) {
+        if (sourceIdentifier.isPrimitive) {
+          const matchedTypes = targetTypes.filter(t => sourceIdentifier.name == t.code);
+          if (matchedTypes.length > 0) {
+            this.markSelectedOptionsInChoice(targetTypes, matchedTypes);
+            continue;
+          }
+          const allowedConvertedTypes = this.findAllowedConversionTargetTypes(sourceIdentifier, targetTypes);
+          if (allowedConvertedTypes.length > 0) {
+            this.markSelectedOptionsInChoice(targetTypes, allowedConvertedTypes);
+            continue;
+          }
+          continue;
+        }
+         // It's a non-primitive source type.  First check if the field is mapped to a BackboneElement.
+        if (targetTypes.length == 1 && targetTypes[0].code == 'BackboneElement') {
+          // TODO: Determine what to do with backbone elements.
+          continue;
+        }
+
+        // Check if the source field has a mapping to a FHIR profile.  If so, and it matches target, apply the profile to the target
+        const originalTargetTypes = common.cloneJSON(targetTypes);
+        const matchedType = this.findMatchingType(sourceValue, targetTypes, sourceIdentifier);
```

Comment:
You changed the signature of `findMatchingType` and `choiceFriendlyEffectiveIdentifier` to support passing in the `sourceIdentifier`.  An alternative solution, that I think is used elsewhere, would be to leave those as-is, and do this instead:
```js
const matchedType = this.findMatchingType(new mdls.IdentifiableValue(sourceIdentifier), targetTypes);
```
I *think* this would accomplish the same goal but without changing the other functions.

Reply:
Wow that is so much cleaner and you are correct it does work the same way. Implemented here https://github.com/standardhealth/shr-fhir-export/pull/161/commits/7139f1c89804af3ba1a5d7a5554617df1512f751.